### PR TITLE
Add secure http handler

### LINKED ISSUE #430: Prevent crashes on http requests to backend
Currently, if an HTTP request to the backend throws an uncaught error, the backend crashes. In production mode, it should instead return a reasonable error to the client and not crash (similar to `secureOn`).

### PR DESCRIPTION
**Pull request by ClFeSc**
Closes #430

### CODE REVIEW

**Comment by ClFeSc** [comment]
I don't think I can do the better typing, so this should be ready for review.

File: backend/src/exercise/http-handler/secure-http.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// TODO: How bad is that?
+// eslint-disable-next-line @typescript-eslint/no-shadow
+import type { Express, Request } from 'express';
```
Comment (Dassderdie) [inline backend/src/exercise/http-handler/secure-http.ts:3]:
```suggestion
import type { Express, Request as ExpressRequest } from 'express';
```
And rename `Request` to `ExpressRequest` in the rest of the file.

File: backend/src/exercise/http-server.ts [hunk 2]
```diff
@@ -26,35 +25,29 @@ export class ExerciseHttpServer {
...
-        app.get('/api/exercise/:exerciseId/history', async (req, res) => {
-            const response = await getExerciseHistory(req.params.exerciseId);
-            res.statusCode = response.statusCode;
-            res.send(response.body);
-        });
+        secureHttp(app, 'get', '/api/health', () => getHealth());
+
+        secureHttp(app, 'post', '/api/exercise', async (req: Request) =>
+            postExercise(databaseService, req.body)
+        );
```
Comment (Dassderdie) [inline backend/src/exercise/http-server.ts:30]: Currently, you replace the whole express API to proxy it. What about only wrapping the handler like this:
```ts
        app.get(
            '/api/health',
            secureHttp((req, res) => {
                const response = getHealth();
                res.statusCode = response.statusCode;
                res.send(response.body);
            })
        );
```

File: backend/src/exercise/http-server.ts [hunk 3]
```diff
@@ -26,35 +23,27 @@ export class ExerciseHttpServer {
...
+
+        app.delete(
+            '/api/exercise/:exerciseId',
+            secureHttp(async (req) => deleteExercise(req.params.exerciseId))
+        );
+
+        app.get(
+            '/api/exercise/:exerciseId/history',
+            secureHttp(async (req) => getExerciseHistory(req.params.exerciseId))
+        );
```
Comment (Dassderdie) [inline backend/src/exercise/http-server.ts:46]:
```suggestion
        app.get('/api/health', (req) => secureHttp(getHealth));

        app.post('/api/exercise', (req) =>
            secureHttp(async () => postExercise(databaseService, req.body))
        );

        app.get('/api/exercise/:exerciseId', (req) =>
            secureHttp(() => getExercise(req.params.exerciseId))
        );

        app.delete('/api/exercise/:exerciseId', (req) =>
            secureHttp(async () => deleteExercise(req.params.exerciseId))
        );

        app.get('/api/exercise/:exerciseId/history', (req) =>
            secureHttp(async () => getExerciseHistory(req.params.exerciseId))
        );
```

File: backend/src/exercise/http-handler/secure-http.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+export function secureHttp<Result extends object | undefined>(
+    operation: (
+        req: ExpressRequest
+    ) => HttpResponse<Result> | Promise<HttpResponse<Result>>
+): (req: ExpressRequest, res: ExpressResponse) => Promise<void> {
+    // TODO: Better type `req` to verify the arguments
+    return async (req: ExpressRequest, res: ExpressResponse) => {
+        let response: HttpResponse<Result>;
+        try {
+            response = await operation(req);
```
Comment (Dassderdie) [inline backend/src/exercise/http-handler/secure-http.ts:25]:
```suggestion
export function secureHttp(
    operation: () =>
        | HttpResponse<object | undefined>
        | Promise<HttpResponse<object | undefined>>
): (req: ExpressRequest, res: ExpressResponse) => Promise<void> {
    return async (_req: ExpressRequest, res: ExpressResponse) => {
        let response: HttpResponse<any>;
        try {
            response = await operation();
```

**Review by Dassderdie — CHANGES_REQUESTED** [review]
I think this is the only way to infer the correct type for `req`...

**Comment by Dassderdie** [comment]
@ClFeSc the pipeline is failing - code wise I'm ok with it.

**Comment by ClFeSc** [comment]
@Dassderdie It should work now...

**Review by Dassderdie — APPROVED** [review]
Still not tested, but code lgtm.

**Pull request merged**